saleor/checkout/steps.py

```diff
@@ -65,156 +125,115 @@ def validate(self):
-
-    def validate(self):
-        super(BillingAddressStep, self).validate()
-        if 'anonymous' in self.forms and not self.anonymous_user_email:
-            raise InvalidData()
+            User.objects.store_address(order.user, self.address, shipping=True)
 
 
-class ShippingStep(BaseAddressStep):
-    template = 'checkout/shipping.html'
-    title = _('Shipping Address')
-    address_form_class = ShippingForm
+class ShippingMethodStep(BaseCheckoutStep):
+    template = 'checkout/shipping_method.html'
+    title = _('Shipping Method')
+    step_name = 'shipping-method'
 
-    def __init__(self, request, storage, cart,
-                 billing_address=None):
-        self.cart = cart
-        address_data = storage.get('address', {})
-        self.billing_address = billing_address
-        if not address_data and billing_address:
-            address = billing_address
-        else:
-            address = Address(**address_data)
-        super(ShippingStep, self).__init__(request, storage, address)
-        delivery_choices = list(
-            (m.name, m) for m in get_delivery_options_for_items(
-                self.cart, address=address))
+    def __init__(self, request, storage, shipping_address, cart, checkout):
+        super(ShippingMethodStep, self).__init__(request, storage, checkout)
+        self.delivery_method = None
         selected_method_name = storage.get('delivery_method')
-        selected_method = None
-        for method_name, method in delivery_choices:
-            if method_name == selected_method_name:
-                selected_method = method
-                break
-        if selected_method is None:
-            # TODO: find cheapest not first
-            selected_method_name, selected_method = delivery_choices[0]
-        self.delivery_method = selected_method
-        self.forms['delivery'] = DeliveryForm(
-            delivery_choices, request.POST or None,
-            initial={'method': selected_method_name})
+        available_deliveries = [
+            {'method': method, 'cost': method.get_delivery_total(cart)}
+            for method in get_delivery_options_for_items(
+                cart, address=shipping_address)]
 
-    def __str__(self):
-        return 'shipping-address'
+        delivery_choices = [(delivery['method'].name, delivery['method'].name)
+                            for delivery in available_deliveries]
 
-    def save(self):
-        delivery_form = self.forms['delivery']
-        if self.forms['address'].cleaned_data.get('shipping_same_as_billing'):
-            address = self.billing_address
+        if available_deliveries:
+            cheapest_delivery = available_deliveries[0]
+
+        for delivery in available_deliveries:
+            if delivery['method'].name == selected_method_name:
+                self.delivery_method = delivery['method']
+                break
+            if delivery['cost'] < cheapest_delivery['cost']:
```

Comment:
Ok, this would be best solution to find cheapest shipping method, but I still need to iterate shipping methods to find selected one, so I think now it is more lazy